onnxruntime/hosting/main.cc

```diff
@@ -1,58 +1,45 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#include "environment.h"
 #include "http_server.h"
+#include "request_handler.h"
 #include "server_configuration.h"
-#include "environment.h"
 
 namespace beast = boost::beast;
 namespace http = beast::http;
-
-void test_request(const std::string& name, const std::string& version,
-                  const std::string& action, onnxruntime::hosting::HttpContext& context) {
-  std::stringstream ss;
-
-  ss << "\tModel Name: " << name << std::endl;
-  ss << "\tModel Version: " << version << std::endl;
-  ss << "\tAction: " << action << std::endl;
-  ss << "\tHTTP method: " << context.request.method() << std::endl;
-
-  http::response<http::string_body>
-      res{std::piecewise_construct, std::make_tuple(ss.str()), std::make_tuple(http::status::ok, context.request.version())};
-
-  res.set(http::field::server, BOOST_BEAST_VERSION_STRING);
-  res.set(http::field::content_type, "plain/text");
-  res.keep_alive(context.request.keep_alive());
-  context.response = res;
-}
+namespace hosting = onnxruntime::hosting;
 
 int main(int argc, char* argv[]) {
-  onnxruntime::hosting::ServerConfiguration config{};
+  hosting::ServerConfiguration config{};
   auto res = config.ParseInput(argc, argv);
 
-  if (res == onnxruntime::hosting::Result::ExitSuccess) {
+  if (res == hosting::Result::ExitSuccess) {
     exit(EXIT_SUCCESS);
-  } else if (res == onnxruntime::hosting::Result::ExitFailure) {
+  } else if (res == hosting::Result::ExitFailure) {
     exit(EXIT_FAILURE);
   }
 
-  onnxruntime::hosting::HostingEnvironment env;
+  hosting::HostingEnvironment env;
   auto logger = env.GetLogger();
 
-  // TODO: below code snippet just trying to show case how to use the "env".
-  //       Will be moved to proper place.
+  // TODO: below code snippet just trying to show case how to use the "env". Move later.
   LOGS(logger, VERBOSE) << "Logging manager initialized.";
   LOGS(logger, VERBOSE) << "Model path: " << config.model_path;
   auto status = env.GetSession()->Load(config.model_path);
   LOGS(logger, VERBOSE) << "Load Model Status: " << status.Code() << " ---- Error: [" << status.ErrorMessage() << "]";
 
   auto const boost_address = boost::asio::ip::make_address(config.address);
 
-  onnxruntime::hosting::App app{};
-  app.Post(R"(/v1/models/([^/:]+)(?:/versions/(\d+))?:(classify|regress|predict))", test_request)
-     .Bind(boost_address, config.http_port)
-     .NumThreads(config.num_http_threads)
-     .Run();
+  hosting::App app{};
+  app.Post(R"(/v1/models/([^/:]+)(?:/versions/(\d+))?:(classify|regress|predict))",
+           [&env](const std::string& name, const std::string& version, const std::string& action, hosting::HttpContext& context) {
```

Comment:
I would prefer to have the main() as short as possible. We may want to move it to a separate "controller" class.

Reply:
What do you mean here? Do you mean move the call to `app.Post` or do you mean the creation of an `App` object to a separate class?

Reply:
Do note: we need to have a lambda here to capture `env` by reference

Reply:
What if you just put the `hosting::Predict(...)` here?

Reply:
In that case, how would I pass `env`?

Reply:
You will anyway need to pass the `env` to the http server (it contains the port, thread count etc.), then you could use it in `HandleRequest()` https://github.com/NonStatic2014/onnxruntime/blob/master/onnxruntime/hosting/http/session.h#L85

Reply:
As we discussed, we could have this change in next PR when you pass through `env` from main to http stack.